solrmonitor/solrmonitor.go

```diff
@@ -311,25 +311,41 @@ func (c *SolrMonitor) shouldWatchChildren(path string) bool {
 }
 
 func (c *SolrMonitor) dataChanged(path string, data string, version int32) error {
-	if !strings.HasPrefix(path, c.solrRoot+"/collections/") || !strings.HasSuffix(path, "/state.json") {
-		return fmt.Errorf("unknown dataChanged: %s", path)
-	}
-	coll := c.getCollFromPath(path)
-	if coll != nil {
-		coll.setData(data, version)
-		c.callSolrListener(coll)
-		if coll.isPRSEnabled() {
-			coll.startMonitoringReplicaStatus()
+	collectionsPrefix := c.solrRoot + "/collections/"
+	if strings.HasPrefix(path, collectionsPrefix) {
+		// we will always need the collection
+		coll := c.getCollFromPath(path)
+		if strings.HasSuffix(path, "/state.json") {
+			// common state.json change
+			if coll != nil {
+				state := coll.setStateData(data, version)
+				c.callSolrListener(coll.name, state)
+				if coll.isPRSEnabled() {
+					coll.startMonitoringReplicaStatus()
+				}
+			}
+			return nil
+		} else if trimmed := strings.TrimPrefix(path, collectionsPrefix); !strings.Contains(trimmed, "/") {
+			// less common (usually just initialization) collection change
+			if coll != nil && coll.name == trimmed {
+				state := coll.setCollectionData(data)
+				if state != nil {
+					// for collection data, we don't hit the callback until the collection state is available as well
+					// on delete, the standard state.json flow will handle passing back the nil state once cleared
+					c.callSolrListener(coll.name, state)
+				}
+			}
+			return nil
```

Comment:
See my comment about this being a semantic change. But also... for code clarity. The indentation could be reduced with early returns.

Reply:
Here is some suggested cleanups:
https://github.com/fullstorydev/gosolr/pull/34/files

Thoughts?

Reply:
That works for me! Will comment there as it definitely looks cleaner but one piece may need an update.